AnkiDroid/src/main/java/com/ichi2/libanki/Sched.java

```diff
@@ -1284,26 +1312,23 @@ private boolean _fillRev() {
             if (lim != 0) {
                 mRevQueue.clear();
                 //don't care about any limit with young cards; just want them all mixed
-                List<Long> maybeYoungCards = getCardIdsWithQuery(String.format("SELECT id " +
-                        "FROM cards " +
-                        "WHERE did = %s " +
-                        "AND queue = 2 " +
-                        "AND due <= %s " +
-                        "AND ivl < 21", did, mToday));
+                List<Long> maybeYoungCards = getCardIdsWithQuery(getCriticalReviewQueryGivenSelect("cards.id", did));
                 if (maybeYoungCards.size() > 0) {
                     mRevQueue.addAll(maybeYoungCards);
                 } else {
-                    List<Long> normalCards = getCardIdsWithQuery(String.format("SELECT id " +
+                    List<Long> normalCards = getCardIdsWithQuery(String.format(
+                            "SELECT DISTINCT cards.id " +
                             "FROM cards " +
+                            "JOIN revlog ON revlog.cid = cards.id AND revlog.factor = cards.factor AND revlog.ivl = cards.ivl " +
                             "WHERE did = %s " +
                             "AND queue = 2 " +
                             "AND due <= %s " +
-                            //first order by relative overdue-ness
-                            "ORDER BY (%s - due * 1.0) / ivl desc, " +
+                            //first order by relative overdue-ness; see getCriticalReviewQueryGivenSelect
+                            "ORDER BY ((%s - due + revlog.ivl) / lastIvl) / ((1.0 * revlog.factor) / 1000) desc, " +
```

Comment:
why the factor of 1000 here? does that convert between units? prevent overflow?

not seeing how it would help in an order by clause.

Reply:
correct, does not change ordering, just here to make logic clearer for readers. Factor (a percentage) is stored in permille instead of percent, hence the division by `1000` and not `100`.